### PR TITLE
Add neo RemoveChannel request

### PR DESCRIPTION
**Pull request by gavin-norman-sociomantic**
Blocked on https://github.com/sociomantic-tsunami/swarm/milestone/27.

### CODE REVIEW

**Comment by gavin-norman-sociomantic** [comment]
A bit WIP. Requires a few doc fixes and more testing.

**Comment by gavin-norman-sociomantic** [comment]
This needs to be reworked to allow RemoveChannel to work on a mirrored channel. See https://github.com/sociomantic-tsunami/dhtproto/pull/61#issuecomment-343895187.

**Comment by gavin-norman-sociomantic** [comment]
Adapted to not include change to Mirror request. Tested with dhtnode (https://github.com/sociomantic-tsunami/dhtnode/pull/43).

**Comment by gavin-norman-sociomantic** [comment]
Updated: added flush call.

**Comment by gavin-norman-sociomantic** [comment]
Updated to use new SupportedCode from swarm v4.5.0.

**Comment by gavin-norman-sociomantic** [comment]
Based on #94. Blocked.

**Comment by gavin-norman-sociomantic** [comment]
Fixed D2 bug.

**Comment by gavin-norman-sociomantic** [comment]
#94 merged. Unblocked.

File: src/dhttest/DhtTestCase.d [hunk 1]
```diff
@@ -214,12 +214,10 @@ abstract class NeoDhtTestCase : TestCase
...
     override public void cleanup ( )
     {
+        this.dht.blocking.removeChannel(this.test_channel);
+
         this.log.info("Shutting down client -- node connection error expected");
         this.dht.neo.shutdown();
-
-        auto legacy_dht = new Legacy.DhtClient;
-        legacy_dht.handshake(10000);
-        legacy_dht.removeChannel(this.test_channel);
```
Comment (nemanja-boric-sociomantic) [inline src/dhttest/DhtTestCase.d:222]: Oh, this is a nice consequence!

File: src/dhttest/cases/neo/GetChannels.d [hunk 2]
```diff
@@ -62,13 +62,10 @@ public class GetChannels : NeoDhtTestCase
...
-        super.cleanup();
-
-        auto legacy_dht = new Legacy.DhtClient;
-        legacy_dht.handshake(10000);
-
         foreach ( channel; this.channels )
-            legacy_dht.removeChannel(channel);
+            this.dht.blocking.removeChannel(channel);
+
+        super.cleanup();
```
Comment (nemanja-boric-sociomantic) [inline src/dhttest/cases/neo/GetChannels.d:68]: Why is this moved after the `removeChannel`?

File: src/dhttest/cases/neo/GetChannels.d [hunk 3]
```diff
@@ -118,12 +113,9 @@ public class GetChannelsRemove : NeoDhtTestCase
...
```
Comment (nemanja-boric-sociomantic) [inline src/dhttest/cases/neo/GetChannels.d:119]: Why is this moved after the removeChannel?

File: relnotes/removechannel.feature.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+## Neo RemoveChannel request
```
Comment (nemanja-boric-sociomantic) [inline relnotes/removechannel.feature.md]: Use new release notes format.

File: src/dhtproto/client/mixins/NeoSupport.d [hunk 5]
```diff
@@ -1430,6 +1466,92 @@ template NeoSupport ( )
...
+                Succeeded,
+                Failed
+            }
+
+            FinishedStatus state;
+
+            void notifier ( Neo.RemoveChannel.Notification info,
+                Neo.RemoveChannel.Args args )
+            {
+                with ( info.Active ) switch ( info.active )
```
Comment (nemanja-boric-sociomantic) [inline src/dhtproto/client/mixins/NeoSupport.d]: `final` switch?

File: src/dhttest/DhtTestCase.d (hunk 1 above)
Comment (gavin-norman-sociomantic) [reply src/dhttest/DhtTestCase.d:222]: Yep!

File: src/dhttest/cases/neo/GetChannels.d (hunk 2 above)
Comment (gavin-norman-sociomantic) [reply src/dhttest/cases/neo/GetChannels.d:68]: Hmm, not sure if there was a reason for that or if it was a mistake. I can try compiling it without that change.

Comment (gavin-norman-sociomantic) [reply src/dhttest/cases/neo/GetChannels.d:68]: Yeah, the reason is that the last thing `super.cleanup` does is shut down the DHT client. Previously, when this test was using a separate client instance to remove the used channels, this had no effect.

Comment (nemanja-boric-sociomantic) [reply src/dhttest/cases/neo/GetChannels.d:68]: Makes sense :+1:

**Comment by gavin-norman-sociomantic** [comment]
Updated:
* Release notes. (Including extra commit fixing Exists notes.)
* `final switch`.

**Comment by gavin-norman-sociomantic** [comment]
Fixed D2 build.

**Comment by nemanja-boric-sociomantic** [comment]
This is red, though, I'm not sure if it's just the matter of a flaky test.

**Comment by gavin-norman-sociomantic** [comment]
Huh, it was green before I rebased.

**Comment by gavin-norman-sociomantic** [comment]
```
[turtle    ] Testing 'Neo Mirror followed by neo Puts and suspend/resume' ...
[turtle    ]     FAIL at ./src/dhttest/cases/neo/Mirror.d:349
[turtle    ]         (expression '1012 == 1014' evaluates to false)
```
I can reproduce this failure locally. The interesting thing is that this only manifests now after this branch was rebased.

**Comment by gavin-norman-sociomantic** [comment]
Looks timing related :( Removing random tests from the test suite makes it pass again.

**Comment by gavin-norman-sociomantic** [comment]
Updated: added another commit that removes flakiness from the Mirror test.

**Comment by gavin-norman-sociomantic** [comment]
Green! :balloon:

File: src/dhtproto/client/request/internal/RemoveChannel.d [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+        Returns:
+            true to continue handling the request (supported); false to abort
+            (unsupported)
+
+    ***************************************************************************/
+
+    private bool handleSupportedCode ( ubyte code )
+    {
+        auto supported = cast(SupportedStatus)code;
+        if ( !RemoveChannel.handleSupportedCodes(supported,
```
Comment (nemanja-boric-sociomantic) [inline src/dhtproto/client/request/internal/RemoveChannel.d]: `return RemoveChannel.handleSupportedCodes(...)`

File: src/dhtproto/client/request/internal/RemoveChannel.d [hunk 7]
```diff
@@ -0,0 +1,324 @@
+/*******************************************************************************
+
+    Client DHT RemoveChannel v0 request handler.
+
+    Copyright:
+        Copyright (c) 2017 sociomantic labs GmbH. All rights reserved.
```
Comment (nemanja-boric-sociomantic) [inline src/dhtproto/client/request/internal/RemoveChannel.d:6]: Copyright 2018.

Comment (nemanja-boric-sociomantic) [reply src/dhtproto/client/request/internal/RemoveChannel.d:6]: (in other files as well)

File: src/dhtproto/node/neo/request/RemoveChannel.d [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        auto ed = this.connection.event_dispatcher();
+
+        // Send status code
+        ed.send(
+            ( ed.Payload payload )
+            {
+                payload.add(this.response);
+            }
+        );
+        ed.flush();
```
Comment (nemanja-boric-sociomantic) [inline src/dhtproto/node/neo/request/RemoveChannel.d:85]: I guess we can leave the `flush` until we move to the new swarm.

Comment (nemanja-boric-sociomantic) [reply src/dhtproto/node/neo/request/RemoveChannel.d:85]: Which reminds me, DHT's requests would benefit from `TCP_NODELAY`, even with the current swarm.

File: src/dhttest/cases/neo/Mirror.d [hunk 9]
```diff
@@ -549,9 +548,8 @@ public class MirrorRemoveChannel : NeoDhtTestCase
                 {
                     case started:
                         // As soon as the mirror is up, remove the channel
-                        this.dht.assign(
-                            this.dht.removeChannel(this.test_channel,
-                                &delNotifier));
+                        this.dht.neo.removeChannel(this.test_channel,
+                            &delNotifier);
```
Comment (nemanja-boric-sociomantic) [inline src/dhttest/cases/neo/Mirror.d:543]: Fun fact: this can be `null` ;-) (not that I would recommend that, I think this way it's clearer).

File: src/dhtproto/client/mixins/NeoSupport.d [hunk 10]
```diff
@@ -455,6 +457,40 @@ template NeoSupport ( )
...
+            auto params = Const!(Internals.RemoveChannel.UserSpecifiedParams)(
+                Const!(RemoveChannel.Args)(channel),
+                Const!(Internals.RemoveChannel.UserSpecifiedParams.SerializedNotifier)(
+                    *(cast(Const!(ubyte[notifier.sizeof])*)&notifier)
+                )
+            );
+
+            auto id = this.assign!(Internals.RemoveChannel)(params);
+            return id;
+        }
```
Comment (nemanja-boric-sociomantic) [inline src/dhtproto/client/mixins/NeoSupport.d:492]: Not sure if it's just the github, but while you're there, it would be nice if this commit would be before adapting the mirror tests to use this method.

File: src/dhtproto/client/mixins/NeoSupport.d [hunk 11]
```diff
@@ -1610,6 +1646,92 @@ template NeoSupport ( )
...
             return res;
         }
+
+        /***********************************************************************
+
+            Struct returned after a RemoveChannel request has finished.
+
+        ***********************************************************************/
+
+        private static struct RemoveChannelResult
```
Comment (nemanja-boric-sociomantic) [inline src/dhtproto/client/mixins/NeoSupport.d]: I forgot, but I think `static` here (in non-function context) is redundant.

File: src/dhtproto/client/mixins/NeoSupport.d [hunk 12]
```diff
@@ -1610,6 +1646,92 @@ template NeoSupport ( )
...
+
+            Returns:
+                RemoveChannelResult struct, indicating the result of the request
+
+        ***********************************************************************/
+
+        public RemoveChannelResult removeChannel ( cstring channel )
+        {
+            auto task = Task.getThis();
+            assert(task, "This method may only be called from inside a Task");
```
Comment (nemanja-boric-sociomantic) [inline src/dhtproto/client/mixins/NeoSupport.d]: `assert(task !is null)`

**Comment by nemanja-boric-sociomantic** [comment]
Is there a way to adapt fakedht to not allow removing channel if the channel is being mirrored (and add appropriate tests)?

**Comment by gavin-norman-sociomantic** [comment]
>Is there a way to adapt fakedht to not allow removing channel if the channel is being mirrored (and add appropriate tests)?

Why?

File: src/dhtproto/client/request/internal/RemoveChannel.d (hunk 7 above)
Comment (gavin-norman-sociomantic) [reply src/dhtproto/client/request/internal/RemoveChannel.d:6]: Well, this code was written in 2017 ;)

**Comment by nemanja-boric-sociomantic** [comment]
Because it's advertised as part of the `RemoveChannel`'s specification, not as the implementation detail:

>  +  * A channel that is being mirrored cannot be removed. 

I do admit it's a nitpick.

File: src/dhtproto/client/request/internal/RemoveChannel.d (hunk 7 above)
Comment (nemanja-boric-sociomantic) [reply src/dhtproto/client/request/internal/RemoveChannel.d:6]: Yeah, I was wondering about that. All right, as long as I have a precedent to refer to in the reviews  I receive ;-)

**Comment by gavin-norman-sociomantic** [comment]
Ahh I see. This is actually outdated. See https://github.com/sociomantic-tsunami/dhtproto/issues/39.

File: src/dhtproto/client/mixins/NeoSupport.d (hunk 10 above)
Comment (gavin-norman-sociomantic) [reply src/dhtproto/client/mixins/NeoSupport.d:492]: Yeah, it is.

File: src/dhtproto/client/mixins/NeoSupport.d (hunk 11 above)
Comment (gavin-norman-sociomantic) [reply src/dhtproto/client/mixins/NeoSupport.d]: Correct :+1:

**Comment by gavin-norman-sociomantic** [comment]
Updated.

File: src/dhtproto/client/request/internal/RemoveChannel.d (hunk 6 above)
Comment (nemanja-boric-sociomantic) [reply src/dhtproto/client/request/internal/RemoveChannel.d]: Ping ;-)

**Review by nemanja-boric-sociomantic — APPROVED** [review]

**Comment by gavin-norman-sociomantic** [comment]
Updated with `handleSupportedCodes` simplification.
